### PR TITLE
Match _heap_free

### PR DESCRIPTION
**Pull request by LightningTH**
Although the while(1) loop is ugly, attempts to make it for() or a conditional while loop failed to generate the same code.

### CODE REVIEW

File: include/functions.h [hunk 1]
```diff
@@ -23,6 +23,7 @@ f32 signF(f32 val);
 
 void* heap_malloc(s32 size);
 void* _heap_malloc(HeapNode* head, u32 size);
+u32 _heap_free(HeapNode* heapNodeList, void* addr_to_free);
```
Comment (ethteck) [inline include/functions.h:26]:
```suggestion
u32 _heap_free(HeapNode* heapNodeList, void* addrToFree);
```

File: src/43F0.c [hunk 2]
```diff
@@ -211,7 +211,74 @@ void* _heap_malloc_tail(HeapNode* head, u32 size) {
     return NULL;
 }
 
-INCLUDE_ASM(s32, "43F0", _heap_free);
+u32 _heap_free(HeapNode* heapNodeList, void* addr_to_free) {
```
Comment (ethteck) [inline src/43F0.c:214]:
```suggestion
u32 _heap_free(HeapNode* heapNodeList, void* addrToFree) {
```

File: src/43F0.c [hunk 3]
```diff
@@ -211,7 +211,74 @@ void* _heap_malloc_tail(HeapNode* head, u32 size) {
...
+    u32 curNodeLength;
+    HeapNode* nextNode;
+    HeapNode* nodeToFreeHeader;
+    HeapNode* tempNode;
+    HeapNode* outNode;
+
+    // if no address to free then return
+    if (!addr_to_free) {
```
Comment (ethteck) [inline src/43F0.c:222]: For pointers, we typically do == NULL

File: src/43F0.c [hunk 4]
```diff
@@ -211,7 +211,74 @@ void* _heap_malloc_tail(HeapNode* head, u32 size) {
...
+        return TRUE;
+    }
+
+    // if we are not allocated then ignore this request
+    nodeToFreeHeader = (HeapNode*)((u8 *)addr_to_free - sizeof(HeapNode));
```
Comment (ethteck) [inline src/43F0.c:227]:
```suggestion
    nodeToFreeHeader = (HeapNode*)((u8*)addr_to_free - sizeof(HeapNode));
```

File: src/heap.c [hunk 5]
```diff
@@ -1,22 +1,22 @@
 #include "common.h"
 
-extern s32 D_80268000;
-extern s32 D_802FB800;
+extern HeapNode* heapCollisionHead; //D_80268000;
```
Comment (ethteck) [inline src/heap.c:3]: We typically name globals with a prefix so they're distinguishable from other variables. so like `heap_collisionHead` for example. and those address comments can be removed

**Review by ethteck — CHANGES_REQUESTED** [review]
Nice job! The goto is unfortunate - I'll see if I can get rid of it later today

File: include/functions.h (hunk 1 above)
Comment (LightningTH) [reply include/functions.h:26]: I should have known better :P

File: src/43F0.c (hunk 3 above)
Comment (LightningTH) [reply src/43F0.c:222]: Done, also altered the other I saw

File: include/functions.h (hunk 1 above)
Comment (ethteck) [reply include/functions.h:26]: all good! :)

File: src/heap.c (hunk 5 above)
Comment (LightningTH) [reply src/heap.c:3]: Changed

File: src/43F0.c (hunk 4 above)
Comment (LightningTH) [reply src/43F0.c:227]: One managed to sneak through

**Comment by BowserSlug** [comment]
🚀 This PR matches 1 function (+0.00%) on `us`.

⚠️ This PR introduces 10 warnings:
- src/316d90.c:46: warning: passing arg 1 of `_heap_create' from incompatible pointer type
- src/43F0.c:392: warning: passing arg 1 of `nuPiReadRom' makes integer from pointer without a cast
- src/43F0.c:396: warning: passing arg 1 of `nuPiReadRom' makes integer from pointer without a cast
- src/43F0.c:407: warning: passing arg 2 of `dma_write_block' makes integer from pointer without a cast
- src/43F0.c:411: warning: passing arg 2 of `dma_write_block' makes integer from pointer without a cast
- src/43F0.c:422: warning: passing arg 1 of `osWritebackDCache' makes pointer from integer without a cast
- src/43F0.c:427: warning: assignment makes pointer from integer without a cast
- src/heap.c:12: warning: passing arg 1 of `_heap_malloc' from incompatible pointer type
- src/heap.c:16: warning: implicit declaration of function `_heap_malloc_tail'
- src/heap.c:20: warning: passing arg 1 of `_heap_free' from incompatible pointer type
- src/heap.c:24: warning: passing arg 1 of `_heap_create' from incompatible pointer type
- src/heap.c:39: warning: passing arg 1 of `_heap_malloc' from incompatible pointer type
- src/heap.c:45: warning: passing arg 1 of `_heap_free' from incompatible pointer type
- src/heap.c:52: warning: passing arg 1 of `_heap_create' from incompatible pointer type
- src/heap.c:5: warning: redundant redeclaration of `heap_battleHead' in same scope
- src/heap.c:60: warning: passing arg 1 of `_heap_malloc' from incompatible pointer type
- src/heap.c:62: warning: passing arg 1 of `_heap_malloc' from incompatible pointer type
- src/heap.c:68: warning: passing arg 1 of `_heap_free' from incompatible pointer type
- src/heap.c:70: warning: passing arg 1 of `_heap_free' from incompatible pointer type
- src/heap.c:8: warning: passing arg 1 of `_heap_create' from incompatible pointer type
- src/world/area_end/end_00/DF6A20.c:57: warning: passing arg 1 of `_heap_free' from incompatible pointer type
- src/world/area_end/end_00/DF6A20.c:57: warning: passing arg 2 of `_heap_free' makes pointer from integer without a cast

File: src/heap.c [hunk 6]
```diff
@@ -1,26 +1,27 @@
 #include "common.h"
 
-extern s32 D_80268000;
-extern s32 D_802FB800;
+extern HeapNode heap_collisionHead; //D_80268000
+extern HeapNode heap_generalHead; //D_802FB800
+extern HeapNode heap_battleHead; //D_803DA800
```
Comment (nanaian) [inline src/heap.c:5]: Remove these old comments

**Review by nanaian — CHANGES_REQUESTED** [review]

File: include/functions.h (hunk 1 above)
Comment (nanaian) [reply include/functions.h:26]: Is this resolved?

Comment (ethteck) [reply include/functions.h:26]: Looks like it's still not renamed here

File: src/heap.c (hunk 6 above)
Comment (LightningTH) [reply src/heap.c:5]: Oops, left them in when doing the mass renaming so I didn't mix them up.

File: include/functions.h (hunk 1 above)
Comment (LightningTH) [reply include/functions.h]: Renamed, thank you for catching functions.h, I got the c file.

File: src/43F0.c [hunk 7]
```diff
@@ -211,7 +211,69 @@ void* _heap_malloc_tail(HeapNode* head, u32 size) {
...
+
+    // see if the next node after us is allocated, if not then adjust our size
+    // to include it and point nextNode to be the node after as it must be allocated
+    if (nextNode && !nextNode->allocated) {
+        curNodeLength += nextNode->length + sizeof(HeapNode);
+        nextNode = nextNode->next;
+    }
+
+    // walk the full heap node list looking for the block before our current entry
+    for (tempNode = heapNodeList;1; tempNode = tempNode->next) {
```
Comment (Brotenko) [inline src/43F0.c:244]:
```suggestion
    for (tempNode = heapNodeList; 1; tempNode = tempNode->next) {
```

Comment (LightningTH) [reply src/43F0.c]: moved to a while loop

**Pull request merged**